### PR TITLE
Add focused benchmark for metric hotpath

### PR DESCRIPTION
**Pull request by cijothomas**
I understand we have a lot of benchmarks, but I wanted to add something very focused, to help with PRs like https://github.com/open-telemetry/opentelemetry-rust/pull/1379 (and its many follow ups).

Also added dedicated one for AttributeSet creation costs for various scenarios.

It okay to nuke these once the perf improvements are done, but I hope this is very useful to quickly measure PRs impact. 
I was able to use this with #1379, and was able to quickly see the improvement.

### CODE REVIEW

**Comment by cijothomas** [comment]
@KallDrexx I used this benchmarks against 1379. And here is my results (~16% boost). Could you check if you see similar?

```
Counter_Add_Sorted      time:   [839.97 ns 853.31 ns 869.54 ns]
                        change: [-16.163% -13.860% -11.672%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  6 (6.00%) high mild
  2 (2.00%) high severe

Counter_Add_Unsorted    time:   [832.49 ns 840.25 ns 848.39 ns]
                        change: [-13.431% -11.739% -10.075%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
```

**Comment by lalitb** [comment]
This is the result of this benchmark test on my machine (~ 20% boost):

for main branch:
```
Counter_Add_Sorted      time:   [1.1975 µs 1.1985 µs 1.1996 µs]
Counter_Add_Unsorted    time:   [1.2020 µs 1.2041 µs 1.2076 µs]
```

for #1379  PR:
```
Counter_Add_Sorted      time:   [959.58 ns 963.14 ns 968.10 ns]
Counter_Add_Unsorted    time:   [965.16 ns 966.00 ns 966.89 ns]
```

File: opentelemetry-sdk/benches/metric_counter.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    let meter = meter_provider.meter("benchmarks");
+    let counter = meter.u64_counter("counter_bench").init();
+    counter
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+    counter_no_attributes(c);
+}
+
+fn counter_no_attributes(c: &mut Criterion) {
```
Comment (lalitb) [inline opentelemetry-sdk/benches/metric_counter.rs:24]: nit - this method is recording measurement with attributes, even though the method name has `no_attributes`.

File: opentelemetry-sdk/benches/metric_counter.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        });
+    });
+
+    c.bench_function("Counter_Add_Unsorted", |b| {
+        b.iter(|| {
+            let mut rng = SmallRng::from_entropy();
+            let index_first_attribute = rng.gen_range(0..2);
+            let index_second_attribute = rng.gen_range(0..5);
+            let index_third_attribute = rng.gen_range(0..10);
+            let index_forth_attribute = rng.gen_range(0..10);
```
Comment (lalitb) [inline opentelemetry-sdk/benches/metric_counter.rs:57]: should we use the same max cardinality for sorted and unsorted for comparison purposes - as of now, unsorted has 1000, and sorted has 1600?

Comment (cijothomas) [reply opentelemetry-sdk/benches/metric_counter.rs:57]: good catch! I was experimenting with few and forgot to fix them to be consistent!

**Review by lalitb — APPROVED** [review]
LGTM.

**Comment by cijothomas** [comment]
Merging now, as to leverage this for the pending PRs. Happy to address any additional feedbacks as followups.

**Pull request merged**